simulation/chart.py

```diff
@@ -1,232 +1,427 @@
+
+    Attributes:
+        width (float): width of the line
+        style (PenStyle): line style
+        color (Color): line color
     """
-    config = {}
 
-    if size is not None:
-        config["symbolSize"] = size
+    width: float
+    style: PenStyle
+    color: Color
 
-    if color is not None:
-        config["symbolBrush"] = color
 
-    if symbol is not None:
-        config["symbol"] = symbol
+# Method to resample the data before plotting ot avoid plotting multiple line segments per pixel.
+DownsampleMethod = Literal[
+    "subsample",  # Downsample by taking the first of N samples. This method is fastest and least accurate.
+    "mean",  # Downsample by taking the mean of N samples.
+    "peak",  # Downsample by drawing a saw wave that follows the min and max of the original data. This method produces the best visual representation of the data but is slower.
+]
 
-    if border is not None:
-        config["symbolPen"] = border
+# Layout defaults
+DEFAULT_WIN_SIZE = (1000, 1000)
+DEFAULT_CODEC = "libx264"
+DEFAULT_PIXEL_FORMAT = "yuv420p"
+DEFAULT_FOREGROUND: Color = "d"
+DEFAULT_BACKGROUND: Color = "k"
+DEFAULT_ANTIALIAS = False
 
-    return chart.plot(name=name, pen=None, **config)
+# Chart defaults
+DEFAULT_CHART_HEIGHT = 200
+DEFAULT_LEGEND_WIDTH = 300
+DEFAULT_AXIS_LOG = False
 
 
-def create_video_process(
-    file_name: str,
-    size: Tuple[int, int],
-    codec: str = "libx264",
-    pixel_format: str = "yuv420p",
-) -> Popen:
-    """Creates a ffmpeg video process which accepts input from stdin.(NA in windows)
+# Plot defaults
+DEFAULT_PEN_COLOR = RGBAColor(200, 200, 200)
+DEFAULT_PEN_WIDTH = 1.0
+DEFAULT_PEN_STYLE = PenStyle.SolidLine
 
-    Args:
-        file_name (str): name of the file to save encoded video ouput
-        size (Tuple[int,int]): Size (width,height) of the video frame
-        codec (str, optional): Codec name to be used for encoding. Defaults to "libx264".
-        pixel_format (str, optional): Ouput pixel format. Defaults to "yuv420p".
+DEFAULT_SYMBOL_SIZE = 10.0
+DEFAULT_SYMBOL_MARKER: SymbolType = "o"
+DEFAULT_SYMBOL_PEN = PenConfig(color=DEFAULT_PEN_COLOR)
+DEFAULT_SYMBOL_COLOR = RGBAColor(50, 50, 150)
+
+DEFAULT_DOWNSAMPLE_METHOD = "peak"
+DEFAULT_DOWNSAMPLE_AMOUNT = 1
+DEFAULT_DOWNSAMMPLE_AUTO = False
 
-    Returns:
-        Process: Asyncronious sub process listening stdin for encoding to the file.
+
+class ChartWidget:
+    """A chart container for single group of plots and a legend.
+
+    Args:
+        chart (PlotItem): base container plotitem
     """
-    process = (
-        ffmpeg.input(
-            "pipe:",
-            format="rawvideo",
-            pix_fmt="rgb24",
-            s=f"{size[0]}x{size[1]}",
+
+    def __init__(self, chart: pg.PlotItem) -> None:
+        self._chart = chart
+        self._plots: Dict[str, pg.PlotDataItem] = {}
+
+    def add_line_plot(
+        self,
+        id: str,
+        name: str,
+        color: Color = DEFAULT_PEN_COLOR,
```

Comment:
Out of curiosity, why are you using globals to set defaults rather than just defining them in the function?

Reply:
I prefer constants which also holds some definition by name.

Reply:
Is there some reasoning beyond a preference? It's not super pythonic to define a bunch of constants instead of using defaults directly. If we're going to deviate from convention, I'd prefer a good reason.

Reply:
This is a preference, updating based on your feedback

Reply:
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/0f379f981088a9a3162a7ae1b81edb5798d4cc52

Reply:
Thanks!